aicsimageio/writers/writer.py

```diff
@@ -1,49 +1,78 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
 from abc import ABC, abstractmethod
+from typing import Any, Optional, Union
 
-from .. import constants, types
+import dask.array as da
+import numpy as np
 
+from ..constants import Dimensions
 
-class Writer(ABC):
-    def __init__(self, file_path: types.PathLike):
-        """
-        Write STCZYX data arrays to file, with accompanying metadata
-        Will overwrite existing files of same name.
+###############################################################################
 
-        Parameters
-        ----------
-        file_path: types.PathLike
-            Path to image output location
 
-        Examples
-        --------
-        Construct and use as object
+class Writer(ABC):
+    """
+    Write a data array to file, with accompanying metadata
+    This will overwrite existing files of same name.
 
-        >>> image = numpy.ndarray([1, 10, 3, 1024, 2048])
-        ... writer = DerivedWriter("file.ome.tif")
-        ... writer.save(image)
-        ... writer.close()
+    Parameters
+    ----------
+    filepath: types.PathLike
+        Path to image output location
 
-        Construct with a context manager
+    Examples
+    --------
+    Construct and use as object
+    >>> image = numpy.ndarray([1, 10, 3, 1024, 2048])
+    ... writer = DerivedWriter("file.ome.tif")
+    ... writer.save(image)
+    ... writer.close()
 
-        >>> image2 = numpy.ndarray([5, 486, 210])
-        ... with DerivedWriter("file2.ome.tif") as writer2:
-        ...     writer2.set_metadata(myMetaData)
-        ...     writer2.save(image2)
-        """
-        self.file_path = file_path
+    Construct with a context manager
+    >>> image2 = numpy.ndarray([5, 486, 210])
+    ... with DerivedWriter("file2.ome.tif") as writer2:
+    ...     writer2.save(image2, metadata=some_metadata_object)
+    """
+    filepath = None
+    _buffer = None
+
+    @abstractmethod
+    def open(self):
+        self._buffer = open(self.filepath, "w")
```

Comment:
Not sure how to handle this and `close`. I think this should be the default but we can also allow custom open and close methods?

Reply:
Should I just remove the `abstractmethod` here?

Reply:
I think @toloudis suggestion makes sense just have save take the file path and then open write and close the file inside save. Other methods just configure how to shape the output.

Reply:
Yeah, I was really trying to discern whether there is a use case for saving a file over multiple separate api calls, in which it makes sense to have some kind of class instance that keeps the file open.  There isn't much need to overengineer things by having a Writer base class when there's no reason to have polymorphic writers and the save interface has custom fields for each writer anyway, that would end up being decoded as kwargs.  I like the simplicity of OmeTiffWriter.save(file, data, metadata...) and CziWriter.save(file, data, metadata...) etc if we can get by on that.

Reply:
See updated PR. Writers are now a single method (`save`) that are `static` and `abstract`.